baybe/campaign.py

```diff
@@ -222,6 +269,84 @@ def add_measurements(
+        self,
+        measurements: pd.DataFrame,
+        numerical_measurements_must_be_within_tolerance: bool,
+    ) -> None:
+        """Mark the given elements of the discrete subspace as measured.
+
+        Args:
+            measurements: A dataframe containing parameter configurations to be
+                marked as measured.
+            numerical_measurements_must_be_within_tolerance: See
+                :func:`baybe.utils.dataframe.fuzzy_row_match`.
+        """
+        idxs_matched = fuzzy_row_match(
+            self.searchspace.discrete.exp_rep,
+            measurements,
+            self.parameters,
+            numerical_measurements_must_be_within_tolerance,
+        )
+        self._searchspace_metadata.loc[idxs_matched, _MEASURED] = True
+
+    @singledispatchmethod
+    def toggle_discrete_candidates(  # noqa: DOC501
+        self,
+        constraints: Collection[DiscreteConstraint] | pd.DataFrame,
+        exclude: bool,
+        complement: bool = False,
+        dry_run: bool = False,
+    ) -> pd.DataFrame:
+        """In-/exclude certain discrete points in/from the candidate set.
+
+        Args:
+            constraints: A filtering mechanism determining the candidates subset to be
+                in-/excluded. Can be either a collection of
+                :class:`~baybe.constraints.base.DiscreteConstraint` or a dataframe.
+                For the latter, see :func:`~baybe.utils.dataframe.filter_df`
+                for details.
+            exclude: If ``True``, the specified candidates are excluded.
+                If ``False``, the candidates are considered for recommendation.
+            complement: If ``True``, the filtering mechanism is inverted so that
+                the complement of the candidate subset specified by the filter is
+                toggled. For details, see :func:`~baybe.utils.dataframe.filter_df`.
+            dry_run: If ``True``, the target subset is only extracted but not
+                affected. If ``False``, the candidate set is updated correspondingly.
+                Useful for setting up the correct filtering mechanism.
+
+        Returns:
+            A new dataframe containing the  discrete candidate set passing through the
+            specified filter.
+        """
+        df = self.searchspace.discrete.exp_rep
+
+        if isinstance(constraints, pd.DataFrame):
+            # Determine the candidate subset to be toggled
+            points = filter_df(df, constraints, complement)
+
+        elif isinstance(constraints, Collection) and is_all_instance(
+            constraints, DiscreteConstraint
+        ):
+            # Filter the search space dataframe according to the given constraint
+            idx = reduce(
+                lambda x, y: x.intersection(y), (c.get_valid(df) for c in constraints)
+            )
+
+            # Determine the candidate subset to be toggled
+            points = df.drop(index=idx) if complement else df.loc[idx].copy()
```

Comment:
Fully agree, that's the part that I mentioned [here](https://github.com/emdgroup/baybe/pull/412#discussion_r1848005887) about the `.filter()` method. That is, the operation should ne be visible to the campaign in the first place but taken care of by the search space, which internally should do pandas/polars or whatever. But let me add a TODO note so that we don't forget about it.

Reply:
done